src/sagemaker/base_serializers.py

```diff
@@ -0,0 +1,396 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""Implements base methods for serializing data for an inference endpoint."""
+from __future__ import absolute_import
+
+import abc
+from collections.abc import Iterable
+import csv
+import io
+import json
+import numpy as np
+from pandas import DataFrame
+from six import with_metaclass
+
+from sagemaker.utils import DeferredError
+
+try:
+    import scipy.sparse
+except ImportError as e:
+    scipy = DeferredError(e)
+
+
+class BaseSerializer(abc.ABC):
+    """Abstract base class for creation of new serializers.
+
+    Provides a skeleton for customization requiring the overriding of the method
+    serialize and the class attribute CONTENT_TYPE.
+    """
+
+    @abc.abstractmethod
+    def serialize(self, data):
+        """Serialize data into the media type specified by CONTENT_TYPE.
+
+        Args:
+            data (object): Data to be serialized.
+
+        Returns:
+            object: Serialized data used for a request.
+        """
+
+    @property
+    @abc.abstractmethod
+    def CONTENT_TYPE(self):
+        """The MIME type of the data sent to the inference endpoint."""
+
+
+class SimpleBaseSerializer(with_metaclass(abc.ABCMeta, BaseSerializer)):
+    """Abstract base class for creation of new serializers.
+
+    This class extends the API of :class:~`sagemaker.serializers.BaseSerializer` with more
+    user-friendly options for setting the Content-Type header, in situations where it can be
+    provided at init and freely updated.
+    """
+
+    def __init__(self, content_type="application/json"):
+        """Initialize a ``SimpleBaseSerializer`` instance.
+
+        Args:
+            content_type (str): The MIME type to signal to the inference endpoint when sending
+            request data (default: "application/json").
+        """
+        super(SimpleBaseSerializer, self).__init__()
+        if not isinstance(content_type, str):
+            raise ValueError(
+                "content_type must be a string specifying the MIME type of the data sent in "
```

Comment:
Technically, we just check if the content type is a string so the error msg returned is misleading